flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractRestHandler.java

```diff
@@ -0,0 +1,201 @@
+				try {
+					request = mapper.readValue("{}", messageHeaders.getRequestClass());
+				} catch (JsonParseException | JsonMappingException je) {
+					log.error("Implementation error: Get request bodies must have a no-argument constructor.", je);
+					sendErrorResponse(new ErrorResponseBody("Internal server error."), HttpResponseStatus.INTERNAL_SERVER_ERROR, ctx, httpRequest);
+					return;
+				}
+			} else {
+				try {
+					ByteBufInputStream in = new ByteBufInputStream(msgContent);
+					request = mapper.readValue(in, messageHeaders.getRequestClass());
+				} catch (JsonParseException | JsonMappingException je) {
+					log.error("Failed to read request.", je);
+					sendErrorResponse(new ErrorResponseBody(String.format("Request did not match expected format %s.", messageHeaders.getRequestClass().getSimpleName())), HttpResponseStatus.BAD_REQUEST, ctx, httpRequest);
+					return;
+				}
+			}
+
+			CompletableFuture<HandlerResponse<P>> response;
+			try {
+				HandlerRequest<R> handlerRequest = new HandlerRequest<>(request, routed.pathParams(), routed.queryParams());
+				response = handleRequest(handlerRequest);
+			} catch (Exception e) {
+				response = FutureUtils.completedExceptionally(e);
+			}
+
+			response.whenComplete((HandlerResponse<P> resp, Throwable error) -> {
+				try {
+					if (error != null) {
+						log.error("Implementation error: Unhandled exception.", error);
+						sendErrorResponse(new ErrorResponseBody("Internal server error."), HttpResponseStatus.INTERNAL_SERVER_ERROR, ctx, httpRequest);
+					} else {
+						if (resp.wasSuccessful()) {
+							sendResponse(messageHeaders, resp.getResponse(), ctx, httpRequest);
+						} else {
+							sendErrorResponse(new ErrorResponseBody(resp.getErrorMessage()), resp.getErrorCode(), ctx, httpRequest);
+						}
+					}
+				} catch (Exception e) {
+					log.error("Critical error while sending a response.", e);
+				}
+			});
+		} catch (Exception e) {
+			log.error("Request processing failed.", e);
+			try {
+				sendErrorResponse(new ErrorResponseBody("Internal server error."), HttpResponseStatus.INTERNAL_SERVER_ERROR, ctx, httpRequest);
+			} catch (IOException e1) {
+				log.error("Critical error while sending a response.", e1);
+			}
+		}
+	}
+
+	/**
+	 * This method is called for every incoming request and returns a {@link HandlerResponse} that either contains a
+	 * {@link ResponseBody} of type {@code P} if the request was handled successfully or an error otherwise.
+	 *
+	 * @param request request that should be handled
+	 * @return future containing a handler response
+	 */
+	protected abstract CompletableFuture<HandlerResponse<P>> handleRequest(@Nonnull HandlerRequest<R> request);
+
+	private static <M extends MessageHeaders<R, P, ?>, R extends RequestBody, P extends ResponseBody> void sendResponse(M messageHeaders, P response, ChannelHandlerContext ctx, HttpRequest httpRequest) throws IOException {
```

Comment:
We don't really need the `messageHeaders` here except for the status code. Wouldn't it make sense to pass only this value to the method?